libglusterfs/src/inode.c

```diff
@@ -176,7 +176,7 @@ hash_dentry(inode_t *parent, const char *name, int mod)
 static int
 hash_gfid(uuid_t uuid, int mod)
 {
-    return ((uuid[15] + (uuid[14] << 8)) % mod);
+    return ((int *)uuid)[1] & (1U >> 1) % mod;
```

Comment:
Any reason for using the second integer instead of the first one ? second integer contains 4 fixed bits. It's not very bad, but it could make the distribution worse.

Also, what's the purpose of `(1U >> 1)` ? It evaluates to 0, so the result of this is always 0.

As an additional improvement, given that `mod` will (should) always be a power of 2, we could just receive a bitmask (= mod - 1) and apply an `and`.

Reply:
No particular reason, ideally we should probably xor mix all 16 bytes down to 32 bits though, so how about:

int* d = (int*)uuid;
int t = d[0] ^ d[1] ^ d[2] ^ d[3];
return t & (mod - 1);

The problem here is that effectively by default we're using every OTHER pair of bytes, but a more comprehensive hash that mixes byte by byte will take relatively longer to compute.  The mixing above will hopefully translate into some or another SSE vector instruction, but I'm not familiar enough with that to know for sure.  And sizeof(int) could vary, so should probably use uint32_t for the variables, return type I'm thinking can remain int.

I was wondering about that 1U >> 1 myself a little earlier, realizing the exact same thing, this will always return 0 - which may explain the timeout on the one test case.

If mod is guaranteed to be a power of 2 then I agree that & (mod - 1) should be much more optimal.

Is there a performance test case that measures the hash table performance (directly or indirectly)?

Reply:
OK, currently we can guarantee it's a power of two, however, if this becomes configurable, may I silently round-up to a power of two?  In general the rule in my training was that hash table sizes should be primes though, not powers of two, but given that the gfid values should be random, I'm comfortable with enforcing a power of 2.

Reply:
> No particular reason, ideally we should probably xor mix all 16 bytes down to 32 bits though, so how about:
> 
> int* d = (int*)uuid; int t = d[0] ^ d[1] ^ d[2] ^ d[3]; return t & (mod - 1);
> 
> The problem here is that effectively by default we're using every OTHER pair of bytes, but a more comprehensive hash that mixes byte by byte will take relatively longer to compute. The mixing above will hopefully translate into some or another SSE vector instruction, but I'm not familiar enough with that to know for sure. And sizeof(int) could vary, so should probably use uint32_t for the variables, return type I'm thinking can remain int.

gfids are already randomly generated, so I would just take a single value. Xor of random data is not more random, so it's unnecessary here. The only thing I commented is that uuids are randomly generated but there are 4 bits that are always fixed to indicate that it's actually a random uuid. These bits fall in the second 4-bytes value, so just taking the first 4-bytes is enough to have 32 fully random bits.

> I was wondering about that 1U >> 1 myself a little earlier, realizing the exact same thing, this will always return 0 - which may explain the timeout on the one test case.
> 
> If mod is guaranteed to be a power of 2 then I agree that & (mod - 1) should be much more optimal.
> 
> Is there a performance test case that measures the hash table performance (directly or indirectly)?

I'm not aware of any specific test doing that. In fact it's hard to test hash table performance inside Gluster because it does a lot of other things. From my point of view this change is fine (except for the `1U >> 1` thing), even if it doesn't provide a visible performance improvement.

> OK, currently we can guarantee it's a power of two, however, if this becomes configurable, may I silently round-up to a power of two? In general the rule in my training was that hash table sizes should be primes though, not powers of two, but given that the gfid values should be random, I'm comfortable with enforcing a power of 2.

If the data is already random, a power of 2 is the best option IMO. Only when data is not known to be random we should use a prime (the dentry cache is using a prime because it depends on a hash of a name, which might not be random enough).

Regarding the silent round-up, I think we have two options:

1. In the description of the option say that it should be a power of 2 and, if it's not, it will be rounded up to the next power of 2, and just do that in the code.
2. In the description of the option say that it must be a power of 2. Then in the code, if the user specifies a non-power of 2, just return an error and don't change the option.

Probably the first one is more user friendly because it doesn't force them to use a calculator (in case they are not used to work with powers of 2).

Reply:
ok, footwork for this is done, and the hash has been updated as per suggestions.